rfcs/text/testing-refinement.md

```diff
@@ -0,0 +1,110 @@
+
+## Visual Comparison testing
+Making assertions of values in the CSS is not the most reliable way to verify the visual styling of the components is correct. In addition, this has been found to be brittle in cases such as where browser upgrades slightly amend the number of pixels of component dimensions and the test spec is asserting on specific pixel values.
+
+The visual comparison testing tool ‘Applitools’ has been under test for some time. It is probably the most mature of this type of tool and is endorsed by prominent members of the testing community. Snapshots of the required screen or frame are taken and compared with baseline snapshots. Any differences that can be detected with the human eye are highlighted as test failures. While we are only currently primarily concerned with testing the components in Chrome, we do have the option with Applitools to render the snapshots of the DOM in almost any other operating system, browser, device and viewport configuration we like by using their Ultrafast Grid. This would give us an early indication of any styling issues with the components in other set ups.
+
+Recently, we have carried out some proof of concept testing of the Carbon components by integrating Applitools with our existing Cypress regression suite. While we can see that this would be a more accurate way of testing the visual attributes of the components, this activity has highlighted that in its current form, the regression suite with Applitools integration is not a viable option as the vast number of comparisons extend an already lengthy suite.
+
+## Reformatting of component stories for testing
+By presenting the various different pre-set configurations of the components in as few frames and stories as possible without the need to set knobs, the tests could run faster while Applitools could take and compare a vastly reduced number of snapshots. The purpose of our Carbon component library is to showcase to Developers how they can create a beautiful UI in a fast and easy manner with our components, not how to amend knobs in Storybook.
+
+# Detailed design
+
+To facilitate a change to this method of regression testing, the components need to be presented differently in Storybook with all possible configurations pre-set in the stories and no knobs to amend. This has already been done to some extent for some components e.g. Button in the `Design System` directory. If all of the configurations and scenarios can be demonstrated in stories in the `Design System` directory then they should, with regression testing run against those stories. Any additional cases, such as edge cases or extended test cases e.g. tests for special characters in labels, should be added in a story in the `Test` directory therefore using the `Test` directory purely for cases for extended testing only.
+
```

Comment:
Would a screenshot be useful here? The Button example where you can see when using a certain icon the left alignment is out would illustrate it well.

Reply:
Yes, I'll add that. At the point I created this, the All Buttons story showing all the button examples including all icons hadn't been created.